### PR TITLE
IRC Federation: RFC2813 implementation (ngIRCd)

### PR DESCRIPTION
**Pull request by alansikora**
### Summary

This is a PR that covers IRC's [RFC2813 implementation](https://tools.ietf.org/html/rfc2813), using a generic approach, ready for expansion (so we can implement TS6 an others).

### How to

Install [ngircd](https://github.com/ngircd/ngircd) and use this very basic [ngircd.conf](https://gist.github.com/alansikora/d0013bb132e6854458ddbb1efad769c4) file.

You can also use `irssi` or any other IRC client to test the other side of the communication.

### Limitations

 - all the usernames get `-irc` and `-rkt` appended to avoid nick conflict problems;
 - the server needs to accept nicknames longer than 9 chars;

### CODE REVIEW

**Comment by geekgonecrazy** [comment]
This looks like it aims to close:
#4667
#9586

Correct? 

Also I see you linked to some examples could you add that to a readme like the previous version?  This way the documentation on how to do it is easier for us to find.

**Comment by geekgonecrazy** [comment]
Can we change the settings on this to make it more clear what exactly the irc here means?

IE its irc federation not just irc support.  Also making clear not the same as the irc bridge etc

**Comment by Hudell** [comment]
Yup. I'm gonna change it to IRC Federation. I also found a few new problems with it that I need to fix before this is merged.

**Comment by bluecmd** [comment]
How functional is this patch? We'd like to give it a try and possibly contribute to it if you need help.
We want to evaluate using RocketChat but we're highly dependent on IRC integration, and federation would be really cool way for us to achieve that - seeing as we already have our own IRC network.

**Comment by Hudell** [comment]
It works for exchanging of messages on Rocket.Chat channels through IRC. I'm currently fixing an issue that is preventing Rocket.Chat from sending messages on the IRC channels.

**Comment by Sing-Li** [comment]
@Hudell  was able to test 
(1) channel existing on RC -> sync'd to IRC    
(2) channel existing on IRC -> sync'd to RC
(3) user existing on RC -> sync'd to IRC with `-rkt` suffix
(4) user existing on IRC -> sync'd to RC with `-irc` suffix

However, if a new channel is created on IRC, the sync to RC fails ... seems there is an `o` appended to the name that makes it fail.  Here is a log for `/join thisroom` where you see the extra `o`

```
I20180522-17:26:16.482(8)? [irc][server] Timeout
I20180522-17:26:38.958(8)? [irc][server] Handling peer message: :joeguy JOIN #thisroomo
I20180522-17:26:38.959(8)? [irc][server] Emitting peer command to local: {"identifier":"joinedChannel","args":{"roomName":"thisroom\u0007o","nick":"joeguy"}}
I20180522-17:26:47.971(8)? [irc][server] Handling peer message: PING :global.rocket.chat
I20180522-17:26:47.972(8)? [irc][server] Sending Command: :irc.rocket.chat PONG irc.rocket.chat
```
Also, if a new channel is created in RC, there is a crash in the server log and then a `#undefined` is often created in IRC.

screen shot 2018-05-22 at 5 30 18 pm

**Comment by evrardjp** [comment]
A welcomed feature.

**Comment by engelgabriel** [comment]
@Hudell can we merge this PR?

**Comment by Hudell** [comment]
Yes

**Comment by evrardjp** [comment]
close PR? :D

File: packages/rocketchat-lib/server/models/Rooms.js [hunk 1]
```diff
@@ -88,6 +88,10 @@ class ModelRooms extends RocketChat.models._Base {
 
 	// FIND
 
+	findWithUsername(username, options) {
```
Comment (sampaiodiego) [inline packages/rocketchat-lib/server/models/Rooms.js]: @rodrigok we'll need to rewrite this on `no-cache` branch

Comment (rodrigok) [reply packages/rocketchat-lib/server/models/Rooms.js]: Yes

**Comment by miwi-fbsd** [comment]
Hi,

Is there anywhere a documention how that should be setup?

**Comment by cutzenfriend** [comment]
Please create a documention on how this works!

**Comment by misse** [comment]
Hi,

Everything seems to be up and running but I'm not getting irc-channels in my channel list. Can I manually try to join an irc-channel via rocketchat somehow?

Documentation / quickstart would be greatly appreciated.

**Comment by Hudell** [comment]
You should see the IRC channels if you click on the search button on Rocket.Chat's sidebar. I'll write a documentation page about using this bridge.

**Comment by misse** [comment]
Hi @Hudell, do you have any update on this? We're not even getting it to connect and we're having a hard time debugging it.

**Comment by jorgeag68** [comment]
hello.

I can see and join channels from both sides. 

But cannot see any messages on either of the sides..

**Comment by dg-otrs** [comment]
Check this comment https://github.com/RocketChat/Rocket.Chat/issues/216#issuecomment-465776108

IRC Federation not working with config from https://github.com/RocketChat/Rocket.Chat/pull/10113#issue-174534240

**Comment by dg-otrs** [comment]
So yes it can connect but no messages will be transferred via the bridge.... its so sad! as this would be a simple solution to get some better accessibility support.

**Pull request merged**